mutt-wizard.1
.\" Automatically generated by Pandoc 2.9.2.1
.\"
.TH "MUTT-WIZARD" "1" "mutt-wizard" "" ""
.hy
.SH NAME
.PP
mutt-wizard - autoconfigure email accounts for neomutt and isync
.SH SYNOPSIS
.PP
\f[B]mutt-wizard\f[R] [ \f[I]OPTIONS\f[R] ]
.SH DESCRIPTION
.PP
\f[B]mutt-wizard\f[R] takes a user email account and sets up a
terminal-based email interface for it with \f[B]neomutt.\f[R] This can
include offline email with \f[B]isync/mbsync\f[R] and configs for
\f[B]msmtp\f[R] for sending mail, and also passwords automatically
encrypted and stored with \f[B]pass.\f[R]
.SH COMMANDS
.TP
\f[B]-a your\[at]email.com\f[R]
add an email address
.TP
\f[B]-l\f[R]
list all email accounts configured by mutt-wizard
.TP
\f[B]-d\f[R]
pick an already configured account and remove its configuration
.TP
\f[B]-D your\[at]email.com\f[R]
remove a configured account without confirmation
.TP
\f[B]-y your\[at]email.com\f[R]
download and upload mail for an email account
.TP
\f[B]-Y\f[R]
sync all email accounts
.TP
\f[B]-t 15\f[R]
toggle a cronjob that syncs your mail every 15 minutes (or any other
number under 60)
.TP
\f[B]-T\f[R]
toggle a cronjob without specifying minutes between sync
.TP
\f[B]-r\f[R]
reorder account shortcut numbers
.TP
\f[B]-c\f[R]
change neomutt theme
.SH OPTIONS FOR ADDING ACCOUNTS
.PP
These can be specified on the command line, otherwise, you will be
prompted for what is necessary.
mutt-wizard knows the IMAP/SMTP server information for most email
providers, so specifying them is usually redundant.
.TP
\f[B]-u billy\f[R]
Account logon/username if required and different from email address.
.TP
\f[B]-n Billy\f[R]
Real name which will appear in emails.
Should be put in quotes if multiple words.
.TP
\f[B]-m number\f[R]
Set a maximum number of messages to be stored offline.
.TP
\f[B]-i\f[R]
IMAP/POP server address
.TP
\f[B]-I\f[R]
IMAP/POP server port (assumed to be 993 for IMAP and 995 for POP if not
specified)
.TP
\f[B]-s\f[R]
SMTP server address
.TP
\f[B]-S\f[R]
SMTP server port (assumed to be 465 if not specified)
.TP
\f[B]-x\f[R]
Account password.
You will be prompted for the password interactively if this option is
not given.
\f[B]-P\f[R] Pass Prefix.
The password will be stored using pass at <passprefix><email>
.SH OTHER OPTIONS
.TP
\f[B]-f\f[R]
Force account creation and guess mailboxes without attempting to connect
to server.
Otherwise if connection cannot be made, the configured account settings
will not be persistent.
.TP
\f[B]-o\f[R]
Create settings for an account to be used online only without mail
syncing abilities.
Without \f[B]-f\f[R] connection will still be attempted in setup to
discover mailboxes.
.TP
\f[B]-X\f[R]
When removing an email profile with either \f[I]-d\f[R] or \f[I]-D,\f[R]
also delete the local mail (will not delete the mail on the server).
.TP
\f[B]-p\f[R]
Use POP protocol instead of IMAP.
Requires \f[I]mpop\f[R] to download mail after configuration.
Server details can still be given with the \f[I]-i\f[R] and \f[I]-I\f[R]
options as if it were a IMAP.
.SH DETAILS
.TP
\f[B]mailsync\f[R]
mutt-wizard calls a script \f[I]mailsync\f[R] to sync mail.
This script additionally indexes new mail with notmuch and gives you a
notification if new mail has arrived.
If you want to bypass its additional features, you can always just run
\f[I]mbsync -a\f[R] to sync your mail directly.
.TP
\f[B]Mail location\f[R]
If the user chooses to keep offline email with \f[B]isync,\f[R] it will
be kept in \f[I]\[ti]/.local/share/mail/.\f[R] \f[B]notmuch\f[R] can be
used to index and search this mail by giving this directory when first
running \f[B]notmuch setup.\f[R] If you have not set up notmuch before,
mutt-wizard will automatically set it up in the background the first
time you add an account.
.TP
\f[B]muttrc files\f[R]
mutt-wizard will create a muttrc file for each created account holding
account-specific details.
These will appear in \f[I]\[ti]/.config/mutt/accounts/\f[R] and can be
edited by the user if needbe.
Note that the mutt-wizard will also source these files and create the
bindings to switch between them, and these will appear in your default
\f[I]\[ti]/.config/mutt/muttrc\f[R] file.
.TP
\f[B]Mail deletion\f[R]
mutt-wizard\[aq]s delete action will delete configuration files and
\f[I]not\f[R] downloaded mail for safety (and time)\[aq]s sake.
If you want to delete downloaded mail, do so manually by removing it
from the directory above.
.TP
\f[B]Default settings\f[R]
The mutt-wizard has many default settings that focus on making it
aesthetically pleasing and supplying more vim-like bindings.
These can be found in
\f[I]/usr/local/share/mutt-wizard/mutt-wizard.muttrc\f[R] and the
default mailcap file can be found in
\f[I]/usr/local/share/mutt-wizard/mailcap.\f[R] Any of these settings
can be overwritten in \f[I]\[ti]/.config/mutt/muttrc,\f[R] but be
mindful that your overriding binds should appear after the
\f[I]mutt-wizard.muttrc\f[R] file is sourced.
.TP
\f[B]Detecting server settings\f[R]
mutt-wizard has a repository of email services and their server
information kept in \f[I]/usr/local/share/mutt-wizard/domains.csv\f[R]
which is used to automatically configure email settings.
If your email provider is not found there, it will prompt you to input
your email service\[aq]s IMAP and SMTP server information which can
usually be found by searching online.
.PP
If you would like to help develop mutt-wizard for others, you are
invited to add this service information to \f[I]domains.csv\f[R] on
mutt-wizard\[aq]s Github <https://github.com/lukesmithxyz/mutt-wizard>
or Gitlab <https://gitlab.com/lukesmithxyz/mutt-wizard> pages.
.TP
\f[B]Gmail accounts\f[R]
Google will require you to allow \[dq]less-secure\[dq] (third party)
applications or remove two-factor authentication in order to access
their IMAP servers to download your mail.
If you use Gmail, be sure to handle this before running mutt-wizard
<https://support.google.com/accounts/answer/6010255>.
.TP
\f[B]Enterprise and university accounts\f[R]
Many universities and businesses might host their domain\[aq]s email via
Google or another service.
This often requires a special IMAP/SMTP-specific password that you must
generate and use.
Again, mutt-wizard can handle these systems, but only once they\[aq]ve
been set up.
.TP
\f[B]Password decryption\f[R]
mutt-wizard uses \f[I]pass\f[R] and therefore \f[I]gpg\f[R] to decrypt
your passwords.
Provided your GPG key has a password, this might mean that you will be
prompted for your GPG password the first time you sync or send mail in a
session.
Once your password is cached, it might also expire later as well.
.PP
Because of this I strongly recommend the program \f[I]pam-gnupg\f[R]
<https://github.com/cruegge/pam-gnupg> which automatically unlocks your
GPG password on login and keeps it active, thus giving you, with
mutt-wizard, secure access to all your email accounts on your system
without ever having to input a password.
.PP
If you don\[aq]t want to use this program, you can also increase the
cache time of an inputted GPG password with the
\f[I]default-cache-ttl\f[R] and \f[I]max-cache-ttl\f[R] variables in
your \f[I]gpg-agent.conf.\f[R]
.SH MUTT-WIZARD\[aq]S NEOMUTT CONFIGURATION
.PP
Here is a list of not only mutt-wizard\[aq]s particular defaults, but
what you need to get the most out of email accounts configured with
mutt-wizard.
.TP
\f[B]Color\f[R]
The mutt-wizard\[aq]s default settings add color to messages in the
index and color mail details to make them easier to see.
New mail, in addition to being marked by the typical N, will also be
bold.
.TP
\f[B]Change theme\f[R]
The mutt-wizard\[aq]s default theme is created by Luke Smith and its
file name is theme-LukeSmith.muttrc in /usr/share/mutt-wizard, To add
new themes, place the file in the same location with the default name as
follows theme-\f[I]name\f[R].muttrc where \f[I]name\f[R] is the name of
theme.
.TP
\f[B]Movement with h/j/k/l\f[R]
Use vim keys to move down \f[I]j\f[R] or up \f[I]k\f[R] in mail, while
\f[I]l\f[R] opens mail, then the attachment view, then an attachment,
while \f[I]h\f[R] is the reverse.
While mail is open, go to next or previous mail with \f[I]J\f[R] and
\f[I]K.\f[R] In the mail index, \f[I]d\f[R] and \f[I]u\f[R] go down and
up by a half page and \f[I]gg\f[R] and \f[I]G\f[R] go to the very top
and very bottom.
.TP
\f[B]Search mail\f[R]
If you have \f[B]notmuch\f[R] configured with your proper mail directory
(see above), you may run \f[I]ctrl-f\f[R] to search for mail containing
any given sequence.
Even without notmuch, \f[I]L\f[R] limits mail, showing only those with
the given sequence in the subject while \f[I]A\f[R] shows all mail (same
as limiting to \[dq]all\[dq]).
.TP
\f[B]Deleting mail\f[R]
\f[I]D\f[R] deletes mail, while \f[I]U\f[R] undeletes it (type in mail
number to get to deleted mail).
Note that \f[I]S\f[R] saves your mailbox, finalizing deletion.
If you have a \f[I]Trash\f[R] box, deleted mail is moved there.
If you want it to skip that and simply be deleted, comment out or remove
the \f[I]set trash\f[R] line in that account\[aq]s muttrc.
.TP
\f[B]Send mail\f[R]
\f[I]m\f[R] creates a new mail message; \f[I]r\f[R] replies to the
selected message; \f[I]R\f[R] replies all to the selected message and
\f[I]f\f[R] forwards the selected message.
.TP
\f[B]Compose mail screen\f[R]
Once you write mail and save the buffer you will be brought to the
compose screen.
Press \f[I]a\f[R] to add attachments, use \f[I]s/t/c/b/d\f[R] to change
the subject/to/CC/BCC/description.
Press \f[I]S\f[R] to change the signature/encryption.
Press \f[I]y\f[R] to send the mail.
.TP
\f[B]Saving and autocompleting email addresses with abook\f[R]
Install the optional dependency abook and you will be able to save the
sender\[aq]s email address with \f[I]a.\f[R] Once this is done, when you
are typing in any email/contact prompt, you may press \f[I]Tab\f[R] to
find contacts matching your input.
Although abook is often used with mutt, it is also a useful program in
its own right.
.TP
\f[B]Switching and moving mail between mailboxes\f[R]
The \f[I]g\f[R] key can be paired with several other keys to
automatically move to another mailbox: gi: Inbox; gs: Sent; gd: Drafts;
ga: Archive; gS: Spam; gj: Junk; gt: Trash.
These bindings will only be present for accounts that have the boxes in
question.
Instead of \f[I]g,\f[R] you can also press \f[I]C\f[R] to copy mail or
\f[I]M\f[R] to move mail to the same boxes.
.TP
\f[B]Switching between accounts\f[R]
mutt-wizard can configure as many as nine accounts each numbered by the
lowest available number when configured.
Press \f[I]i\f[R] followed by an account\[aq]s number to change to that
account: i2, i5, etc.
\f[I]ctrl-b\f[R] to open a menu to select a url you want to open in you
browser.
.TP
\f[B]Sidebar\f[R]
mutt-wizard enables the sidebar by default which displays your
account\[aq]s boxes with mail tallies.
\f[I]B\f[R] will toggle the sidebar.
Move up and down in it with \f[I]ctrl-k/j.\f[R] Open a box with
\f[I]ctrl-o.\f[R]
.TP
\f[B]More information\f[R]
Remember that you can press \f[I]?\f[R] at any time in neomutt to get a
list of all key-bindings and functions.
This list can also vary for different context menus.
.SH AUTHORS
.PP
Written by Luke Smith <luke\[at]lukesmith.xyz> originally in 2018.
.SH LICENSE
.PP
GPLv3
.SH SEE ALSO
.PP
\f[B]mailsync\f[R](1), \f[B]neomutt\f[R](1), \f[B]neomuttrc\f[R](1)
\f[B]mbsync\f[R](1), \f[B]mpop\f[R](1), \f[B]msmtp\f[R](1),
\f[B]notmuch\f[R](1), \f[B]abook\f[R](1)


changelog.txt
                History (newest first)

Generate man page from readme, to avoid duplicate descriptions.
Readme in rst instead of md.
Separate changelog.txt.

Add tests and made bug fixes

*urlscan* instead of *urlview*

*wm* generates *mutt* config on every full sync,
-  to reflect changes in mailboxes in the shortcuts
-  to reflect changes in *mbsync/getmail* config in *mutt* config

*wm* integrates *mailsync*,
not to overload the system namespace and because of code reuse

*remove|rm* instead of *delete*, *list|ls* instead of only *ls*

Make channel name equal to email address
to avoid choosing a new name for the same thing.

Other more vim-like shortcut changes

*gm/gM* to sync mail inside *mutt*, as *o/O* has a *mutt* assignment already.

Honor *$MAILDIR*, *$XDG_CONFIG_HOME*, *$XDG_CACHE_HOME*, if defined.

`isync`/`mbsync` has replaced `offlineimap` as the backend.
Offlineimap was error-prone, bloated, used obsolete Python 2 modules and required separate steps to install the system.

`mw` is now an installed program instead of just a script needed to be kept in your mutt folder.

`dialog` is no longer used (less bloat) and the interface is simply text commands.

More autogenerated shortcuts that allow quickly moving and copying mail between boxes.

More elegant attachment handling.

Image/video/pdf attachments without relying on the neomutt instance.

abook integration by default.

The messy template files and other directories have been moved or removed, leaving a clean config folder.

msmtp configs moved to `~/.config/` and mail default location moved to `~/Mail`, reducing mess in `~`.

`pass` is used as a password manager instead of separately saving passwords.

Script is POSIX sh compliant.

Addition of a manual `man mw`.


mutt-wizard.sh
#!/usr/bin/env bash

[ "$(uname)" == "Darwin" ] && os=".macos"

muttdir="$HOME/.config/mutt"
namere="^[a-z_][a-z0-9_-]*$"
emailre=".+@.+\..+"

createMailboxes() { \
	tmpdir=$(mktemp -d)
	offlineimap --info -a "$1" 2&> "$tmpdir"/log
	sed -n '/^Folderlist/,/^Folderlist/p' "$tmpdir"/log |
		grep "^ " | sed -e "s/\//./g;s/(.*//g;s/^ //g" > "$tmpdir"/lognew
	while read box; do mkdir -p "$HOME/.mail/$1/$box"; done <"$tmpdir"/lognew ;}

testSync() { (crontab -l | grep mailsync.sh && removeSync) || addSync ;}

addSync() { min=$(dialog --inputbox "How many minutes should be between mail syncs?" 8 60 3>&1 1>&2 2>&3 3>&-)
	(crontab -l; echo "*/$min * * * * eval \"export $(egrep -z DBUS_SESSION_BUS_ADDRESS /proc/$(pgrep -u $LOGNAME -x i3)/environ)\"; "$muttdir"/etc/mailsync.sh") | crontab - &&
	dialog --msgbox "Cronjob successfully added. Remember you may need to restart or tell systemd/etc. to start your cron manager for this to take effect." 7 60 ;}

removeSync() { ((crontab -l | sed -e '/mailsync.sh/d') | crontab - >/dev/null) && dialog --msgbox "Cronjob successfully removed. To reactivate, select this option again." 6 60 ;}

changePassword() { \
	gpgemail=$( dialog --title "Luke's mutt/offlineIMAP password wizard" --inputbox "Insert the email address with which you originally created your GPG key pair. This is NOT necessarily the email you want to configure." 10 60 3>&1 1>&2 2>&3 3>&- )
	while ! [[ "${gpgemail}" =~ ${emailre} ]]; do
		gpgemail=$(dialog --no-cancel --title "Luke's mutt/offlineIMAP autoconfig" --inputbox "That's not a valid email address. Please input the entire address." 10 60 3>&1 1>&2 2>&3 3>&1)
	done
	dialog --title "Luke's mutt/offlineIMAP password wizard" --passwordbox "Enter the new password for the \"$1\" account." 10 60 2> /tmp/$1
	gpg2 -r "$gpgemail" --encrypt /tmp/$1 || (dialog --title "GPG decryption failed." --msgbox "GPG decryption failed. This is either because you do not have a GPG key pair or because your distro uses GPG1 and you thus need to symlink /usr/bin/gpg2 to /usr/bin/gpg." 7 60 && break)
	shred -u /tmp/$1
	mv /tmp/$1.gpg "$muttdir"/credentials/
	dialog --title "Finalizing your account." --infobox "The account \"$title\"'s password has been changed. Now attempting to configure mail directories...

	This may take several seconds..." 10 70
	createMailboxes "$title" || (clear && exit)
	detectMailboxes "$title"
	dialog --title "Password changed." --msgbox "Your \"$fulladdr\" password has been changed. To start the download of your mail, you can manually run \`offlineimap -a $title\` in a terminal. The first sync may take some time depending on the amount of your mail." 8 60 ;}

chooseDetect() { for x in $(grep "^accounts =" ~/.offlineimaprc | sed -e 's/accounts =\( \)//g;s/\(,\) /\n/g;'); do detectMailboxes $x; done && detectSuccess ;}

detectWarning() { \
	dialog --title "Mailbox detect requirement" --yesno "In order for the mailbox detection system to work, you must have

1) already have added the email account with this wizard, and
2) already have run offlineimap at least once to synchronize your mail.

This detection system only works if you already have begun your first sync. If you have, press 'yes' to continue." 12 70 ;}

detectSuccess() { dialog --title "Mailbox detect complete." --msgbox "The script has now smartly detected your different mailboxes and has enabled them in the sidebar and given you keyboard shortcuts as below:

gi for move to the Inbox.
gs for move to Sent Mail.
gd for move to Drafts
gS for move to Spam
ga for move to the Archive.
gj for move to Junk
gt for move to Trash

These shortcuts will only work if your email system does have that particular folder (i.e. if your email system has a Junk folder, but not a Trash folder, 'gt' will not work, etc." 20 60 ;}

formatShortcut() { \
	while read data; do
	echo "macro index,pager g$1 \"<change-folder>$data<enter>\" \"Go to $2.\"" >> "$muttdir"/accounts/$3.muttrc
	done ;}

gen_delim() { \
	delim="="
	for i in `seq $(( $1 - 1 ))`
	do
		delim="$delim-"
	done
	echo $delim ;}

detectMailboxes() { \
	find ~/.mail/$1 -maxdepth 1 -mindepth 1 -type d | sed -e "s/.*\///g;s/^/=/g" > /tmp/$1_boxes
	sidebar_width=$(sed -n -e '/^set sidebar_width/p' "$muttdir"/muttrc | awk -F'=' '{print $2}')
	delim=$(gen_delim $sidebar_width)
	oneline=$(sed -e "s/^\|$/\"/g" /tmp/$1_boxes | tr "\n" " ")
	oneline="=$1 $delim $oneline"
	sed -i "/^mailboxes\|^set record\|^set postponed\|^set trash\|^set spoolfile/d" "$muttdir"/accounts/$1.muttrc
	echo mailboxes $oneline >> "$muttdir"/accounts/$1.muttrc
	sed -i "/^macro index,pager g/d" "$muttdir"/accounts/$1.muttrc
	grep -i /tmp/$1_boxes -e inbox | sed 1q | formatShortcut i inbox $1
	grep -i /tmp/$1_boxes -e sent | sed 1q | formatShortcut s sent $1
	grep -i /tmp/$1_boxes -e draft | sed 1q | formatShortcut d drafts $1
	grep -i /tmp/$1_boxes -e trash | sed 1q | formatShortcut t trash $1
	grep -i /tmp/$1_boxes -e spam | sed 1q | formatShortcut S spam $1
	grep -i /tmp/$1_boxes -e junk | sed 1q | formatShortcut j junk $1
	grep -i /tmp/$1_boxes -e archive | sed 1q | formatShortcut a archive $1
	spoolfile=$(grep -i /tmp/$1_boxes -e inbox | sed -e 's/=/+/g' | sed 1q)
	record=$(grep -i /tmp/$1_boxes -e sent | sed -e 's/=/+/g' | sed 1q)
	postponed=$(grep -i /tmp/$1_boxes -e draft | sed -e 's/=/+/g' | sed 1q)
	trash=$(grep -i /tmp/$1_boxes -e trash | sed -e 's/=/+/g' | sed 1q)
	echo "set spoolfile = \"$spoolfile\"" >> "$muttdir"/accounts/$1.muttrc
	echo "set record = \"$record\"" >> "$muttdir"/accounts/$1.muttrc
	echo "set postponed = \"$postponed\"" >> "$muttdir"/accounts/$1.muttrc
	echo "set trash = \"$trash\"" >> "$muttdir"/accounts/$1.muttrc ;}

# Get all accounts in ~/.offlineimaprc and load into variable `accounts`.
getAccounts() { \
	grep "^accounts =" ~/.offlineimaprc | sed -e 's/accounts =\( \)//g;s/\(,\) /\n/g;' | nl --number-format=ln > /tmp/numbered
	accounts=()
	while read n s ; do
		accounts+=($n "$s" off)
	done < /tmp/numbered ;}

# Yields a menu of available accounts.
inventory() { \
	getAccounts
	choices=$(dialog --separate-output --checklist "Select all desired email accounts with <SPACE>." 15 40 16 "${accounts[@]}" 2>&1 >/dev/tty)

	if [ -z "$choices" ];
		then
			clear
		else
		userchoices=$(IFS="|"; keys="${choices[*]}"; keys="${keys//|/\\|}"; grep -w "${keys}" /tmp/numbered  | awk '{print $2}')
	fi ;}


removeAccount() { sed -ie "
	/Account $1]/,/Account/{//!d}
	/Account $1]/d
	s/ $1\(,\|$\)//g
	s/=$1\(,\|$\)/=/g
	s/,$//g
	" ~/.offlineimaprc
	rm "$muttdir"/accounts/$1.muttrc
	rm "$muttdir"/credentials/$1.gpg
	rm -rf "$muttdir"/accounts/$1
	sed -i "/$1.muttrc/d" "$muttdir"/personal.muttrc 
	# Delete from the line matching the account name, until the next account or empty line
	sed -i "/account $1/,/^\(\s*$\|account\)/d" ~/.msmtprc ;}

manual() { \
	imap=$( dialog --inputbox "Insert the IMAP server for your email provider (excluding the port number)" 10 60 3>&1 1>&2 2>&3 3>&- )
	iport=$(dialog --inputbox "What is your server's IMAP port number? (Usually 993)" 10 60 3>&1 1>&2 2>&3 3>&-)
	smtp=$( dialog --inputbox "Insert the SMTP server for your email provider (excluding the port number)" 10 60 3>&1 1>&2 2>&3 3>&- )
	sport=$( dialog --inputbox "What is your server's SMTP port number? (Usually 587 or 465)" 10 60 3>&1 1>&2 2>&3 3>&- ) ;}


addloop() { fulladdr=$( dialog --title "Luke's mutt/offlineIMAP autoconfig" --inputbox "Insert the full email address for the account you want to configure." 10 60 3>&1 1>&2 2>&3 3>&- )
while ! [[ "${fulladdr}" =~ ${emailre} ]]; do
	fulladdr=$(dialog --no-cancel --title "Luke's mutt/offlineIMAP autoconfig" --inputbox "That's not a valid email address. Please input the entire address." 10 60 3>&1 1>&2 2>&3 3>&1)
done
# Check to see if domain is in domain list
serverinfo=$(grep -w ^${fulladdr##*@} "$muttdir"/autoconf/domains.csv)
if [ -z "$serverinfo" ];
	then
		manual
	else
# Read in server data as variables
IFS=, read service imap iport smtp sport <<EOF
$serverinfo
EOF
fi
realname=$( dialog --title "Luke's mutt/offlineIMAP autoconfig" --inputbox "Enter the full name you'd like to be identified by on this email account." 10 60 3>&1 1>&2 2>&3 3>&- )
title=$(dialog --title "Luke's mutt/offlineIMAP autoconfig" --inputbox "Give a short, one-word name for this email account that will differentiate it from other email accounts." 10 60 3>&1 1>&2 2>&3 3>&1) || exit
while ! [[ "${title}" =~ ${namere} ]]; do
	title=$(dialog --no-cancel --title "Luke's mutt/offlineIMAP autoconfig" --inputbox "Account title not valid. Give a username beginning with a letter, with only lowercase letters, - or _." 10 60 3>&1 1>&2 2>&3 3>&1)
done
login=$(dialog --title "Luke's mutt/offlineIMAP autoconfig" --inputbox "If you have a username for the \"$title\" account which is different from your email address, enter it here. Otherwise leave this prompt blank." 10 60 3>&1 1>&2 2>&3 3>&- )
# Sets the repo type and other variables for the sed regex.
if [[ "$service" == "gmail.com" ]];
	then
		type="Gmail"
		delet="remotehost"
	else
		type="IMAP"
		delet="Gmail]\/"
fi
if [[ -z "$login" ]];
	then
		login=$fulladdr
fi
# The replacements
replacement="
	s/\$realname/$realname/g;
	s/\$title/$title/g;
	s/\$fulladdr/$fulladdr/g;
	s/\$imap/$imap/g;
	s/\$iport/$iport/g;
	s/\$smtp/$smtp/g;
	s/\$sport/$sport/g;
	s/\$type/$type/g;
	s/\$login/$login/g;
	/$delet/d"
# Gets the first unused shortcut number in the muttrc and puts it in $idnum.
grep "i[0-9]" "$muttdir/personal.muttrc" | awk '{print $3}' | sed -e 's/i//g' > /tmp/mutt_used
echo -e "1\\n2\\n3\\n4\\n5\\n6\\n7\\n8\\n9" > /tmp/mutt_all_possible
idnum=$(diff /tmp/mutt_all_possible /tmp/mutt_used | sed -n 2p | awk '{print $2}')
addAccount \
;}

addAccount() {
	# First, adding the encrypted password.
	dialog --title "Luke's mutt/offlineIMAP password wizard" --passwordbox "Enter the password for the \"$title\" account." 10 60 2> /tmp/$title
	gpg2 -r $gpgemail --encrypt /tmp/$title || (dialog --title "GPG decryption failed." --msgbox "GPG decryption failed. This is either because you do not have a GPG key pair or because your distro uses GPG1 and you thus need to symlink /usr/bin/gpg2 to /usr/bin/gpg." 7 60 && break)
	shred -u "/tmp/$title"
	mv "/tmp/$title.gpg" "$muttdir"/credentials/

	# Adding directory structure for cache.
	mkdir -p "$muttdir/accounts/$title/cache/bodies"

	# Creating the offlineimaprc if it doesn't exist already.
	if [ ! -f ~/.offlineimaprc ]; then cp "$muttdir"/autoconf/offlineimap_header"$os" ~/.offlineimaprc; fi
	sed -e "$replacement" "$muttdir"/autoconf/offlineimap_profile"$os" >> ~/.offlineimaprc
	mkdir -p "$HOME/.mail/$title"

	# Creating msmtprc if it doesn't exist already.
	if [ ! -f ~/.msmtprc ]; then cp "$muttdir"/autoconf/msmtprc_header ~/.msmtprc; fi
	sed -e "$replacement" "$muttdir"/autoconf/msmtprc_profile >> ~/.msmtprc

	# Add the mutt profile.
	sed -e "$replacement" "$muttdir"/autoconf/mutt_profile > "$muttdir/accounts/$title.muttrc"
	# Add a numbered shortcut in the muttrc
	echo "macro index,pager i$idnum '<sync-mailbox><enter-command>source \"$muttdir\"/accounts/$title.muttrc<enter><change-folder>!<enter>;<check-stats>'" >> "$muttdir"/personal.muttrc

	# Add to offlineimaprc sync list.
	sed -i.bu "s/^accounts =.*[a-zA-Z]$/&, $title/g;s/^accounts =\\s*$/accounts = $title/g" ~/.offlineimaprc && rm ~/.offlineimaprc.bu

	# Makes account default if there is no default account.
	grep "$muttdir"/personal.muttrc -e "^source .*accounts.*" >/dev/null && echo there || \
	echo "source ${muttdir}/accounts/$title.muttrc" >> "$muttdir"/personal.muttrc

	dialog --title "Finalizing your account." --infobox "The account \"$title\" has been added. Now attempting to configure mail directories...

	This may take several seconds..." 10 70
	createMailboxes "$title" || (clear && exit)
	detectMailboxes "$title"
	dialog --title "Account added." --msgbox "Your \"$fulladdr\" account has been added. To start the download of your mail, you can manually run \`offlineimap -a $title\` in a terminal. The first sync may take some time depending on the amount of your mail." 8 60 ;}

# This is run when a user chooses to add an account.
chooseAdd() { \
	mkdir -p "$muttdir"/credentials/ "$muttdir"/accounts/
	gpgemail=$( dialog --title "Luke's mutt/offlineIMAP password wizard" --inputbox "Insert the email address with which you originally created your GPG key pair. This is NOT necessarily the email you want to configure." 10 60 3>&1 1>&2 2>&3 3>&- )
	addloop
	while : ;
	do
		dialog --title "Luke's mutt/offlineIMAP password wizard" --yesno "Would you like to add another email account?" 5 60 || break
		addloop
	done ;}

wipe () { rm "$HOME/.offlineimaprc"
	rm -rf "$muttdir"/accounts
	rm -f "$muttdir"/credentials/*gpg
	rm "$muttdir"/personal.muttrc ;}

while : ;
	do
choice=$(dialog --title "Luke's mutt/offlineIMAP wizard" --nocancel \
	--menu "What would you like to do?" 15 45 8 \
	0 "List all email accounts configured." \
	1 "Add an email account." \
	2 "Enable/disable autosync." \
	3 "Redetect mailboxes." \
	4 "Change an account's password." \
	5 "Remove an email account." \
	6 "Remove all email accounts." \
	7 "Exit this wizard." \
	 3>&1 1>&2 2>&3 3>&1 )

case $choice in
0) dialog --title "Accounts detected" --msgbox "The following accounts have been detected:
$(grep ~/.offlineimaprc -e "^accounts =" | sed 's/accounts =//g')
" 6 60;;
1) chooseAdd;;
2) testSync;;
3) detectWarning && chooseDetect;;
4) inventory && for i in $userchoices; do changePassword "$i" ; done;;
5) inventory && for i in $userchoices; do removeAccount "$i" ; done;;
6) (dialog --defaultno --title "Wipe all custom neomutt/offlineIMAP settings?" --yesno "Would you like to wipe all of the mutt/offlineIMAP settings generated by the system?" 6 60 && wipe) ;;
7) clear && break ;;
*) echo "Unable to read response from dialog. Exiting." >&2; exit 2
esac
done
